pkgs/development/interpreters/php/generic.nix

```diff
@@ -337,6 +337,7 @@ let
             maintainers = teams.php.members;
             platforms = platforms.all;
             outputsToInstall = [ "out" "dev" ];
+            broken = (versionOlder version "8.1") && (isInOldestRelease 2305); # Will automatically break php < 8.0 with 23.11
```

Comment:
As discussed in #224505, I do not think this is correct. PHP itself is not broken.

Reply:
Why not? I thought PHP 8.0 is going EOL and therefore no longer supported by us come November?

Reply:
PHP 8.0 will be removed completely by then (preparing the PR to do that is in my todo list for next week).

What purpose does it serve to mark it as broken now?

Reply:
I wouldn't flat out remove in but rather mark it as broken/insecure. That way you can still opt into using dependant packages on your own risk.

The maintenance burden for a (marked broken) php80 should be close to nil.

Reply:
At the same time, nixpkgs isn't a collection of unmaintained or insecure software. Sooner or later it will become a huge maintenance burden when some dependency is updated that the old version of PHP won't support. Should we then override and pin older versions of everything there?

So no, we'll drop the old version that will become EOL from unstable early in the release cycle so we have time to find out what's broken and attempt to patch it before the next release, or just mark it as broken by then during the ZHF process.

Reply:
> Sooner or later it will become a huge maintenance burden when some dependency is updated that the old version of PHP won't support.

At that point, you'd simply remove that old, deprecated version of PHP. It will have likely been marked as broken/insecure for a while by then, so nobody can complain they weren't warned.

As you correctly stated, we're not a collection of unmaintained or insecure software. No support is guaranteed here. What I'm concerned about however is to keep those deprecated versions around until they actually cause real issues. As long as they do not cause real issues, they do not hurt us.

I wouldn't advocate for keeping it if there were no users but in this case we do have significant users within Nixpkgs. We're even using limesurvey right now for [this year's community survey](https://discourse.nixos.org/t/nix-community-survey-2023/28988).

> just mark it as broken by then during the ZHF process

This is effectively what my PR does. On branch-off, php80 will "automatically" be marked as broken. No 23.11 user can build a system with php80 without explicitly allowing the package.

Reply:
The fact we were not able to remove PHP 8.0 entirely for 23.05 is already less than ideal: PHP 8.0 security supports ends a bit before the sunset of 23.05.

I'm still not sure to understand what's the point of marking PHP 8.0 as broken right now? It is likely to confuse users as they will end up compiling the package instead of pulling it from the binary cache and there is no easy way to understand why. PHP 8.0 can still be built and installed with no issues which is what `meta.broken` is about.

> I wouldn't advocate for keeping it if there were no users but in this case we do have significant users within Nixpkgs. We're even using limesurvey right now for [this year's community survey](https://discourse.nixos.org/t/nix-community-survey-2023/28988).

PHP 8.1 was released 1.5 year ago, if upstream cannot keep up with PHP releases resources are probably best spent on working to help them with the compat issues instead of trying to keep it in a somewhat working state in `nixpkgs` that will ultimately be insecure.

For Limesurvey since bc48fa8f5e3c09b1c401b21388f2f429df1ad536 the 6.x version was released and it supports PHP 8.1.

https://manual.limesurvey.org/LimeSurvey_roadmap#LimeSurvey_6.0
https://manual.limesurvey.org/index.php?title=Installation_-_LimeSurvey_CE&type=revision&diff=176140&oldid=175637

Reply:
> The fact we were not able to remove PHP 8.0 entirely for 23.05 is already less than ideal: PHP 8.0 security supports ends a bit before the sunset of 23.05.

Agreed.

> I'm still not sure to understand what's the point of marking PHP 8.0 as broken right now?

Again, it won't mark it as broken just yet. It will be marked broken once the next release maintainer bumps the version before branch-off which, in this case, should be right around the time it goes EOL anyways.

> It is likely to confuse users as they will end up compiling the package instead of pulling it from the binary cache and there is no easy way to understand why.

Agreed. Though until we have [a better way to do this](https://github.com/NixOS/rfcs/pull/127), this is how it's done unfortunately.

Another way to do it would be to keep it unbroken but start adding CVEs as they come in; marking the package as insecure. I don't really like this option because it'd communicate to the user that php80 is still supported and the expectation of supported software on a stable release is to stay supported, which it won't. Also maintenance burden.

An alternative that I have thought of/discovered just now would be to treat it like openssl_1_1; add a pseudo vulnerability: https://github.com/NixOS/nixpkgs/blob/c66b67fde8b00849790cd5d5fb68d2a1cc2d860e/pkgs/development/libraries/openssl/default.nix#L236-L238

Since there's precedent for this in Nixpkgs, would you be fine with that?

> PHP 8.1 was released 1.5 year ago, if upstream cannot keep up with PHP releases resources are probably best spent on working to help them with the compat issues instead

This is the best solution. I wish it was an option. Nobody should use or support EOL software.

[Grocy upstream is not very cooperative and does not see any issue.](https://github.com/grocy/grocy/issues/2251)

They've got 8.1 support in their dev branch but I wouldn't be comfortable shipping it. I have no idea about PHP, so I couldn't do a backport either.

> For Limesurvey since [bc48fa8](https://github.com/NixOS/nixpkgs/commit/bc48fa8f5e3c09b1c401b21388f2f429df1ad536) the 6.x version was released and it supports PHP 8.1.

Great!